solr/benchmark/src/java/org/apache/solr/bench/search/NumericSearch.java

```diff
@@ -0,0 +1,298 @@
+      QueryRequest req = new QueryRequest(q);
+      req.setBasePath(basePath);
+      QueryResponse response = req.process(miniClusterState.client, COLLECTION);
+      Set<String> numbers =
+          response.getFacetField("numbers_i_dv").getValues().stream()
+              .map(FacetField.Count::getName)
+              .collect(Collectors.toSet());
+      Set<String> lowCardinalityStrings =
+          response.getFacetField("term_low_s").getValues().stream()
+              .map(FacetField.Count::getName)
+              .collect(Collectors.toSet());
+      Set<String> highCardinalityStrings =
+          response.getFacetField("term_high_s").getValues().stream()
+              .map(FacetField.Count::getName)
+              .collect(Collectors.toSet());
+
+      lowCardTerms = new CircularIterator<>(lowCardinalityStrings);
+      highCardTerms = new CircularIterator<>(highCardinalityStrings);
+      queries = new CircularIterator<>(buildSetQueries(numbers));
+    }
+
+    private Set<String> buildSetQueries(Set<String> numbers) {
+      Iterator<String> valueIterator = numbers.iterator();
+      Set<String> setQueries = new HashSet<>();
+      StringBuilder builder = new StringBuilder();
+      while (valueIterator.hasNext()) {
+        for (int i = 0; i < setQuerySize; i++) {
+          if (!valueIterator.hasNext()) {
+            break;
+          }
+          builder.append(valueIterator.next()).append(",");
+        }
+        builder.setLength(builder.length() - 1);
+        setQueries.add(builder.toString());
+        builder.setLength(0);
+      }
+      return setQueries;
+    }
+
+    @Setup(Level.Iteration)
+    public void setupIteration(MiniClusterState.MiniClusterBenchState miniClusterState)
+        throws SolrServerException, IOException {
+      // Reload the collection/core to drop existing caches
+      CollectionAdminRequest.Reload reload = CollectionAdminRequest.reloadCollection(COLLECTION);
+      reload.setBasePath(miniClusterState.nodes.get(0));
+      miniClusterState.client.request(reload);
+    }
+
+    public QueryRequest intSetQuery(boolean dvs) {
+      return setQuery("numbers_i" + (dvs ? "_dv" : ""));
+    }
+
+    public QueryRequest longSetQuery(boolean dvs) {
+      return setQuery("numbers_l" + (dvs ? "_dv" : ""));
+    }
+
+    public QueryRequest doubleSetQuery(boolean dvs) {
+      return setQuery("numbers_d" + (dvs ? "_dv" : ""));
+    }
+
+    public QueryRequest floatSetQuery(boolean dvs) {
+      return setQuery("numbers_f" + (dvs ? "_dv" : ""));
+    }
+
+    QueryRequest setQuery(String field) {
+      QueryRequest q =
+          new QueryRequest(
+              new SolrQuery(
+                  "q",
+                  termQueryField + ":" + lowCardTerms.next(),
+                  "fq",
+                  "{!terms cache=false f='" + field + "'}" + queries.next()));
+      q.setBasePath(basePath);
+      return q;
+    }
+
+    private static class CircularIterator<T> implements Iterator<T> {
```

Comment:
minor suggestion. flavors of this exist already in commons-collections. you could consider replacing this with something like `org.apache.commons.collections4.iterators.LoopingIterator` or `org.apache.commons.collections4.iterators.LoopingListIterator`.

Reply:
While the benchmark is currently using just one thread, we could increase it to test different things, and those two aren't thread safe. I'd rather keep the one we have here for now